### ISSUE TITLE
indentation of list() of functions

### ISSUE THREAD

**Issue by mmaechler** [emacs-ess/ESS#279]
kind of a deja-vu ... but of course different.  Thanks to our great regression test collection:
This currently indents as
```{r}
a <- list(ex1 = function(p) {
    cMah <- qchisq(0.975, p)   
    function(d) as.numeric(d < cMah)
},
ex2 = function(p) {
    cM <- qchisq(0.95, p)
    function(d) as.numeric(d < cM)
})
```
where notably the  `},`  alignment to column 0 is very bad (and the one of `ex2 = ...` is just a consequence of that.

**Comment by lionel-** [comment]
I believe that indentation is on purpose. It makes most sense with higher order functionals like `lapply()` where you provide a function, and with DSL functions like `test_that()` where you supply blocks of code to a call.

But even here, where would you place the closing `}`? In the same column as the elements of the function body? That wouldn't make sense.

Though if you set `ess-offset-block` to `prev-call` and `ess-indent-from-lhs` to `nil`, you can have this alignment instead:
```r
a <- list(param1 = function() {
         body1
     },
     param2 = function() {
         body2
     })
```
But it will probably look bad with other patterns. So if it's important, we'll need to introduce additional flexibility to set different offset types for different calls (higher order functionals vs containers, etc)

**Comment by lionel-** [comment]
You can also have complete alignment, à la if-else, by adding `fun-decl` to the list `ess-align-blocks`.

Then you have:
```r
a <- list(param1 = function() {
                       body1
                   },
          param2 = function() {
                       body2
                   })
```
but I find it a bit wacky.

**Comment by mmaechler** [comment]
Hmm,  I'm surprised that you find the current default acceptable.
Good programmeRs will not unrarely work with `list(.)`s of small  functions.
Indenting these list components  all the way to column 0 seems plain wrong to me.

I would argue that something between the two cases (boundaries possibly included in this convex combination :-)  you show above is desirable.
My personal preference  is also in between the two but closer to the second:
```{r}
a <- list(param1 = function() {
                 body1
             },
             param2 = function() {
                 body2
             })
```

**Comment by lionel-** [comment]
> Hmm, I'm surprised that you find the current default acceptable.

We discussed these issues at length with @vspinu. The challenge is to limit the complexity of the interface and of the code, while providing a style that works well for most cases. I find the current behaviour not only acceptable but even desirable, because it allows this kind of formatting:
```r
object <- lapply(object, function(x) {
    body
})

test_that("this test works", {
    body
})
```
I believe that a majority of our users want to indent code this way too. With your scheme we'd have things like
```r
object <- lapply(object, function(x) {
                             body
                         })
```

That said, I wouldn't be necessarily opposed to try and offer a more flexible interface. I've been doing a bit of C/C++ programming lately, and cc-mode has a very flexible indenting scheme where you can specify different types of offset for far more semantic units than we currently offer. I currently don't have time to work on that kind of new features though.

**Comment by vspinu** [comment]
> We discussed these issues at length with @vspinu. 

Indeed. I believe this one was the issue that started the whole rewrite. R studio does the same shift to the left in such situations. There is really no other way if we want to save space by shifting function bodies to the left. 

> Good programmeRs will not unrarely work with list(.)

To paraphrase, "functional programmers use nested calls a lot" and it's very unpleasant when you reach 80th column with a couple of such calls.   

I still think that the space saving default is a good tradeoff as most of the usages will concern only one functional parameter.

One can alleviate the problem a bit by shifting the functions on a new line:

```
some_value <-
    list(param1 =
             function() {
                 body1
             },
         param2 =
             function() {
                 body2
             })

some_value <-
    list(param1 = function() {
        body1
    },
    param2 = function() {
        body2
    })
```

**Comment by mmaechler** [comment]
Well,  indentation to column 0 in such a case is just not acceptable to me.

**Comment by lionel-** [comment]
> Well, indentation to column 0 in such a case is just not acceptable to me.

But what do you want to do about it? What kind of setting and UI?

**Comment by vspinu** [comment]
> is just not acceptable to me.

You can always change the setting then. This has been discussed at length and there is little that could be done about it without breaking other places.

**Comment by lionel-** [comment]
personnally I think the best style for list of functions is the following (and it is how it currently works with both RRR and RStudio styles):
```r
object <- list(
  param1 = function(x) {
    body1
  },
  param2 = function(x) {
    body2
  }
)
```

**Comment by vspinu** [comment]
> personnally I think the best style for list of functions 

Oh. Of course. I also had this style in mind when I suggested breaking the line. I just broke it in wrong places.

**Comment by mmaechler** [comment]
I don't disagree that there are alternative styles via different line breaking.
But let me get back to the issue:  **The indentation is just  plain "wrong" **  and you are comparisons to the `lapply()` {or much less convincing `testthat`} are not to the point:

If we have
```{r}
obj <- funcall(...
    ....
    ....
    ....
)
```
we have one single expression which is only complete with the final `)`.   {{R would print the continuation prompt  `+ `  if entered interactively}}.
Consequently  **NO** indentation should go back to column 0  but the final `)`.
That's the  "universal"  way of indenting incomplete expressions.

**Comment by vspinu** [comment]
Before we get into same old discussion please see #164 and #169. 

We have been indenting function bodies to the left for many years and users find it very convenient (#164). R-studio does the same. So, it just cannot be "plain wrong". Saying that `lapply`, which is arguably the most used pattern in R, is less convincing than an obscure `list of functions` pattern is an exaggeration.  Especially if there is a simple "workaround" of splitting the first line in right place. 

If you have a concrete solution (other than changing the default style to less greedy one) we are all ears but, as the aforementioned issues suggest, this might not be that easy.

**Comment by lionel-** [comment]
ooh by the way, you can also have `(setq ess-offset-block 'open-delim)` in your config, after which indentation goes exactly as you want it:
```r
a <- list(param1 = function() {
              body1
          },
          param2 = function() {
              body2
          })

object <- lapply(object, function(x) {
                     body
                 })
```

I would definitely not make this the default though, because your "universal" way of indenting would upset many users. But that setting is included in RRR+ so you can simply have `(ess-set-style 'RRR+)` in your init file.

Now if you really want to introduce more flexibility than we currently have, please come up with a UI design first, and then we can discuss it.

**Comment by lionel-** [comment]
about this RRR+ config Vitalie, I found out that the term "hanging" is really commonplace after all ;)

http://www.gnu.org/software/emacs/manual/html_node/ccmode/Hanging-Braces.html

**Comment by mmaechler** [comment]
@vspinu : It seems I was not understandable about `lapply()`:   That case *is* fine the way it is now,
**but** that is not an argument for the  `a <- list(..)` case:
The `lapply` case only indents to column zero once the expression is *complete* ..
whereas this is not at all true for my case.

**Comment by lionel-** [comment]
> The lapply case only indents to column zero once the expression is complete

That is not true. Here's what happens when you specify additional arguments to the lambda:
```r
object <- lapply(object, function(x, ...) {
  body
},
further_argument)
```
Arguments are pushed from their natural indentation so they never hang. I.e., this is how the engine indented before the rewrite:
```r
object <- lapply(object, function(x, ...) {
  body
},
                 further_argument)
```
Pushing the further args to the left is the only sane option here.

**Comment by lionel-** [comment]
Note that from a practical point of view, you would arrange the further argument as follow in real code:
```r
object <- lapply(object, function(x, ...) {
  body
}, further_argument)
```
or you would take advantage of parameter matching in one way:
```r
object <- lapply(object, param = further_argument, function(x) {
  body
})
```
or another:
```r
object <- lapply(object, further_argument, FUN = function(x) {
  body
})
```
or you'd just give a name to the lambda and put the whole thing on a one-liner.

**Comment by lionel-** [comment]
here's a backward-compatible interface that we could use to introduce flexibility:
```lisp
(setq ess-offset-block '((open-delim :calls ("list" "lst") ;; call names inlined
                         (prev-line :calls some-var))) ;; some-var contains list of call names
                         (prev-call :default)) ;; default offset type
```
The indentation code would shoot errors if the list doesn't contain one plist with `:default` to specify the default behaviour. We could also have a function `(ess-set-offset 'block offset-list)` that would check consistency and would be the preferred way of setting an offset.

For backward-compatibility we'd allow:
```lisp
(setq ess-offset-block 'prev-line))    ;; shortcut for '((prev-line :default))
(setq ess-offset-block '(prev-line 4)) ;; shortcut for '((prev-line :offset 4 :default))
```

**Comment by vspinu** [comment]
> here's a backward-compatible interface that we could use to introduce flexibility:

It's not bad, but then how will it integrate with other variables? Right now we have more or less consistent interface. I am afraid this will be broken with such additions.

**Comment by lionel-** [comment]
I think it should be ok. Anyway, I wouldn't work on that until a long while, so we still have time to think about it.

**Comment by lionel-** [comment]
Since the three offset types and the three semantic units (arguments, arguments-newline and blocks) already interact very well with each other, I don't think there should be any problem as long as it's clear which offset applies in which function call. This should be unambiguous if the setting is well formed (i.e., doesn't include twice the same call name for different offset type).

**Comment by mmaechler** [comment]
@lionel- : Thank you for the clarification about the `obj <- lapply({ .....})`  case.
You are right indeed, in the case of the extra argument.  I still find the result ugly, also for `lapply(.)`  as an incomplete expression should not indent back to column 0 in such cases.  I want to see immediately that the expression is incomplete.

But we do have to agree that we disagree about this issue quite a bit.  A matter of taste more than
anything else.

BTW: there's also the small issue #278  which seems to have remained unnoticed.

**Comment by lionel-** [comment]
So you'd like to have both
```r
a <- list(param1 = function() {
              body1
          },
          param2 = function() {
              body2
          })
```
and
```r
a <- list(param1 = function() {
    body1
})
```
depending on the number of blocks in a function call?

**Comment by lionel-** [comment]
and / or only if the block is the last argument?

The rules would be complicated if we added a setting for this.
